src/R/Support/Impl/Help/IntelliSenseRHost.cs

```diff
@@ -0,0 +1,61 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.ComponentModel.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Common.Core.Shell;
+using Microsoft.Languages.Editor.Shell;
+using Microsoft.R.Host.Client;
+using Microsoft.R.Support.Settings;
+
+namespace Microsoft.R.Support.Help {
+    [Export(typeof(IIntellisenseRHost))]
+    public sealed class IntelliSenseRHost : IIntellisenseRHost {
+        private static readonly Guid SessionId = new Guid("8BEF9C06-39DC-4A64-B7F3-0C68353362C9");
+        private readonly ICoreShell _coreShell;
+        private readonly IRSessionProvider _sessionProvider;
+        private readonly SemaphoreSlim _sessionSemaphore = new SemaphoreSlim(1, 1);
+
+        [ImportingConstructor]
+        public IntelliSenseRHost(ICoreShell coreShell, IRSessionProvider sessionProvider) {
```

Comment:
How taxing is it, actually? If the main problem is that it's delaying other evals, then you could make your evals reentrant, like package manager does for its package fetching - then they can be interrupted by other stuff. But it will block the REPL prompt. So if we're talking hundreds of milliseconds long, it's an option. If it's seconds, we need a separate session.